client/core/reserves.go

```diff
@@ -0,0 +1,171 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+package core
+
+import (
+	"sync/atomic"
+
+	"decred.org/dcrdex/client/asset"
+)
+
+type reservesCounter struct {
+	done uint32 // mutable atomic
+
+	num  uint32 // max lot/swap count
+	mult uint64 // amount reserved per swap
+	rem  uint64 // ideally zero, but handled if not
+}
+
+func newReservesCounter(num uint32, total uint64) *reservesCounter {
+	return &reservesCounter{
+		num:  num,
+		mult: total / uint64(num),
+		rem:  total % uint64(num),
+	}
+}
+
+func (rc *reservesCounter) releaseN(count uint32) uint64 {
+	if rc == nil {
+		return 0
+	}
+	newDone := atomic.AddUint32(&rc.done, count)
+	if newDone > rc.num { // caller bug, but work with it
+		prevDone := newDone - count
+		if prevDone >= rc.num {
+			return 0
+		}
+		count = rc.num - prevDone
+	}
+
+	unlock := uint64(count) * rc.mult
+	if newDone >= rc.num { // final
+		unlock += rc.rem
+	}
+	return unlock
+}
+
+func (rc *reservesCounter) releaseAll() uint64 {
+	if rc == nil {
+		return 0
+	}
+	prevDone := atomic.SwapUint32(&rc.done, rc.num)
+	if prevDone >= rc.num { // releaseAll will be routinely called when an order is retired
+		return 0
+	}
+	count := rc.num - prevDone
+	return uint64(count)*rc.mult + rc.rem
+}
+
+// accountRedeemer is equivalent to calling
+// xcWallet.Wallet.(asset.RedeemReserver) on the to-wallet.
+func (t *trackedTrade) accountRedeemer() (asset.RedeemReserver, bool) {
+	ar, is := t.wallets.toWallet.Wallet.(asset.RedeemReserver)
+	return ar, is
+}
+
+func (t *trackedTrade) reserveRedeems(count uint32) { // once
+	if count == 0 {
+		return
+	}
+	redeemer, is := t.accountRedeemer()
+	if !is {
+		return
+	}
+	total, err := redeemer.ReserveNRedemptions(uint64(count), t.wallets.toAsset)
+	if err != nil {
+		t.dc.log.Errorf("error reserving for %d redemptions: %v", count, err)
+		return // that's alright?
```

Comment:
I think this error should be returned. This could mean that the wallet doesn't have enough to pay for the redemption tx fees.